<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Objectives" data-toc-modified-id="Objectives-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Objectives</a></span></li><li><span><a href="#Example-Together" data-toc-modified-id="Example-Together-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Example Together</a></span><ul class="toc-item"><li><span><a href="#Question" data-toc-modified-id="Question-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Question</a></span></li><li><span><a href="#Considerations" data-toc-modified-id="Considerations-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Considerations</a></span></li><li><span><a href="#Loading-the-Data" data-toc-modified-id="Loading-the-Data-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Loading the Data</a></span></li><li><span><a href="#Some-Exploration-to-Better-Understand-our-Data" data-toc-modified-id="Some-Exploration-to-Better-Understand-our-Data-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Some Exploration to Better Understand our Data</a></span></li><li><span><a href="#Experimental-Setup" data-toc-modified-id="Experimental-Setup-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Experimental Setup</a></span><ul class="toc-item"><li><span><a href="#What-Test-Would-Make-Sense?" data-toc-modified-id="What-Test-Would-Make-Sense?-2.5.1"><span class="toc-item-num">2.5.1&nbsp;&nbsp;</span>What Test Would Make Sense?</a></span></li><li><span><a href="#The-Hypotheses" data-toc-modified-id="The-Hypotheses-2.5.2"><span class="toc-item-num">2.5.2&nbsp;&nbsp;</span>The Hypotheses</a></span></li><li><span><a href="#Setting-a-Threshold" data-toc-modified-id="Setting-a-Threshold-2.5.3"><span class="toc-item-num">2.5.3&nbsp;&nbsp;</span>Setting a Threshold</a></span></li></ul></li><li><span><a href="#$\chi^2$-Test" data-toc-modified-id="$\chi^2$-Test-2.6"><span class="toc-item-num">2.6&nbsp;&nbsp;</span>$\chi^2$ Test</a></span><ul class="toc-item"><li><span><a href="#Setup-the-Data" data-toc-modified-id="Setup-the-Data-2.6.1"><span class="toc-item-num">2.6.1&nbsp;&nbsp;</span>Setup the Data</a></span></li><li><span><a href="#Calculation" data-toc-modified-id="Calculation-2.6.2"><span class="toc-item-num">2.6.2&nbsp;&nbsp;</span>Calculation</a></span></li></ul></li><li><span><a href="#Interpretation" data-toc-modified-id="Interpretation-2.7"><span class="toc-item-num">2.7&nbsp;&nbsp;</span>Interpretation</a></span></li></ul></li><li><span><a href="#Exercise" data-toc-modified-id="Exercise-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Exercise</a></span></li></ul></div>

In [1]:
import numpy as np
import pandas as pd

from scipy import stats
import seaborn as sns

# Objectives

- Conduct an A/B test in Python
- Interpret the results of the A/B tests for a stakeholder

# Example Together

## Question

We have data about whether customers completed sales transactions, segregated by the type of ad banners to which the customers were exposed.

The question we want to answer is whether there was any difference in sales "conversions" between desktop customers who saw the sneakers banner and desktop customers who saw the accessories banner in the month of May 2019.

## Considerations

What would we need to consider when designing our experiment?

Might include:

- Who is it that we're including in our test?
- How big of an effect would make it "worth" us seeing?
    - This can affect sample size
    - This can give context of a statistically significant result
- Other biases or "gotchas"

## Loading the Data

First let's download the data from [kaggle](https://www.kaggle.com/podsyp/how-to-do-product-analytics) via the release page of this repo: https://github.com/flatiron-school/ds-ab_testing/releases 

The code below will load it into our DataFrame:

In [2]:
# This will download the data from online so it can take
# some time (but relatively small download)

df = pd.read_csv('https://github.com/flatiron-school/ds-ab_testing/\
releases/download/v1.2/products_small.csv')

> Let's take a look while we're at it

In [3]:
df.head()

Unnamed: 0                          order_id  \
0     4122928  3e6c5e89fdddcaee0eed210ec2c9cadf   
1      564306  feed6203517d3abf6aab13761633174b   
2     1872289  e33d5d7941edc281646aa37763729771   
3     3616779  7c4caa8d508fa7c3bbc25f35cdd9168a   
4     5871482  12874b29bde8bbd43fb2b95735caf9e6   

                            user_id                           page_id  \
0  90d58d967eb72656e86059ec6f208092  2fdc16a09e0016555dd4da4a3fe84414   
1  08703dab1f004eabba25aacb7f0e5484  6b0a902b9b73d5a158d0119d6feb38ac   
2  bdf1d25697e21419901c94fabdafad15  9ddb7315c4357929931b48f2b3d11c62   
3  8d2f23a732c9527d95678088a3bac122  1f86cd0bea31d54a5b511b42fd19401a   
4  5a22604f8f31ae98ee1211ece3a02004  b533f7e2003418c63fd71471264c559a   

       product site_version                 time        title  target  
0  accessories      desktop  2019-03-06 08:42:47  banner_show       0  
1     sneakers       mobile  2019-04-19 18:30:45  banner_show       0  
2      company       mobile  2019-01-20 17:20:10  banner_show       0  
3     sneakers       mobile  2019-02-20 09:38:32  banner_show       0  
4     sneakers       mobile  2019-04-24 09:19:02  banner_show       0

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000000 entries, 0 to 999999
Data columns (total 9 columns):
 #   Column        Non-Null Count    Dtype 
---  ------        --------------    ----- 
 0   Unnamed: 0    1000000 non-null  int64 
 1   order_id      1000000 non-null  object
 2   user_id       1000000 non-null  object
 3   page_id       1000000 non-null  object
 4   product       1000000 non-null  object
 5   site_version  1000000 non-null  object
 6   time          1000000 non-null  object
 7   title         1000000 non-null  object
 8   target        1000000 non-null  int64 
dtypes: int64(2), object(7)
memory usage: 68.7+ MB


## Some Exploration to Better Understand our Data

Lets's look at the different banner types:

In [5]:
df['product'].value_counts()

clothes             210996
company             203020
sneakers            201298
sports_nutrition    193200
accessories         191486
Name: product, dtype: int64

In [6]:
df.groupby('product')['target'].value_counts()

product           target
accessories       0         186127
                  1           5359
clothes           0         197720
                  1          13276
company           0         203020
sneakers          0         193375
                  1           7923
sports_nutrition  0         190363
                  1           2837
Name: target, dtype: int64

Let's look at the range of time-stamps on these data:

In [7]:
df['time'].min()

'2019-01-01 00:00:25'

In [8]:
df['time'].max()

'2019-05-31 23:59:21'

Let's check the counts of the different site_version values:

In [9]:
df['site_version'].value_counts()

mobile     718521
desktop    281479
Name: site_version, dtype: int64

In [10]:
df['title'].value_counts()

banner_show     872275
banner_click     98330
order            29395
Name: title, dtype: int64

In [11]:
len(df[df['title'] == 'order']), sum(df['target'])

(29395, 29395)

In [12]:
df.groupby('title').agg({'target': 'mean'})

target
title               
banner_click       0
banner_show        0
order              1

## Experimental Setup

We need to filter by site_version, time, and product:

In [13]:
df_AB = df[(df['site_version'] == 'desktop') &
           (df['time'] >= '2019-05-01') &
           ((df['product'] == 'accessories') | (df['product'] == 'sneakers'))]\
.reset_index(drop=True)

In [14]:
df_AB.tail()

Unnamed: 0                          order_id  \
25967     5738290  ecfe9da2dbea255f05b2f526a4c663a7   
25968     7885931  8c28b45ff7fbb5d5df0b7670183689db   
25969     8099155  c613db2f19c7285f101196062c178490   
25970     6690689  6f7f32ecf6790e4eb06384b479f43162   
25971     7650805  af3a19fcceb477d2992f1771af3a9b34   

                                user_id                           page_id  \
25967  cd7598ba5798ad6bd3f5a667ae4c217a  644ad50dd900d6735d8ceb15b71d50f8   
25968  2f28106a4db2babcfde1c4095c805632  a3d2de7675556553a5f08e4c88d2c228   
25969  97473368941b89c4716c4f389e9b8bd7  a3d2de7675556553a5f08e4c88d2c228   
25970  73f403b4843c70c41c0db294ee479582  13f395d7c86ef276f7ce52a557fd3491   
25971  515cb969971fb5e08ae0316f58e3f434  ef38d906f15db1efde867c68f1336946   

           product site_version                 time        title  target  
25967  accessories      desktop  2019-05-01 23:01:05  banner_show       0  
25968     sneakers      desktop  2019-05-26 11:19:09        order       1  
25969     sneakers      desktop  2019-05-11 09:04:46        order       1  
25970  accessories      desktop  2019-05-09 09:17:24  banner_show       0  
25971     sneakers      desktop  2019-05-20 17:58:16  banner_show       0

### What Test Would Make Sense?

Since we're comparing the frequency of conversions of customers who saw the "sneakers" banner against those who saw the "accessories" banner, we can use a $\chi^2$ test.

Note there are other hypothesis tests we can use but this should be fine since it should fit our criteria.

### The Hypotheses

$H_0$: Customers who saw the sneakers banner were no more or less likely to buy than customers who saw the accessories banner.

$H_1$: Customers who saw the sneakers banner were more or less likely to buy than customers who saw the accessories banner.

### Setting a Threshold

We'll set a false-positive rate of $\alpha = 0.05$.

## $\chi^2$ Test

### Setup the Data

We need our contingency table: the numbers of people who did or did not submit orders, both for the accessories banner and the sneakers banner. 

In [15]:
df_A = df_AB[df_AB['product'] == 'accessories']
df_B = df_AB[df_AB['product'] == 'sneakers']

In [16]:
accessories_orders = sum(df_A['target'])
sneakers_orders = sum(df_B['target'])

accessories_orders, sneakers_orders

(496, 799)

To get the numbers of people who didn't submit orders, we get the total number of people who were shown banners and then subtract the numbers of people who did make orders.

In [17]:
accessories_total = len(df_A)
sneakers_total = len(df_B)

accessories_no_orders = accessories_total - accessories_orders
sneakers_no_orders = sneakers_total - sneakers_orders

accessories_no_orders, sneakers_no_orders

(12207, 12470)

In [18]:
contingency_table = np.array([
    (accessories_orders, accessories_no_orders), 
    (sneakers_orders, sneakers_no_orders)
])

contingency_table

array([[  496, 12207],
       [  799, 12470]])

### Calculation


In [19]:
stats.chi2_contingency(contingency_table)

(60.94640624666488,
 5.86498985882812e-15,
 1,
 array([[  633.38922686, 12069.61077314],
        [  661.61077314, 12607.38922686]]))

This extremely low $p$-value suggests that these two groups are genuinely performing differently. In particular, the desktop customers who saw the sneakers banner in May 2019 bought at a higher rate than the desktop customers who saw the accessories banner in May 2019.

## Interpretation

In [20]:
contingency_table

array([[  496, 12207],
       [  799, 12470]])

In [21]:
# Find the difference in conversion rate
accessory_CR, sneaker_CR = contingency_table[:, 0] / contingency_table.sum(axis=1)

In [22]:
print(f'Conversion Rate for accessory banner:\n\t{100*accessory_CR:.3f}%')
print(f'Conversion Rate for sneaker banner:\n\t{100*sneaker_CR:.3f}%')
print('')
print(f'Absolute difference of CR: {100*(sneaker_CR-accessory_CR):.3f}%')

Conversion Rate for accessory banner:
	3.905%
Conversion Rate for sneaker banner:
	6.022%

Absolute difference of CR: 2.117%


So we can say:
- There was a statistically significant difference at the $\alpha$-level (confidence level)
- The difference was about $2.12\%$ in favor of the sneaker banner!

# Exercise

> The company is impressed with what you found and is now wondering if there is a difference in their other banner ads!

With your group, look at the same month (May 2019) but compare different platforms ('mobile' vs 'desktop') and or different banner types ('accessories', 'sneakers', 'clothes', 'sports_nutrition'). Just don't repeat the same test we did above 😉

Make sure you record what considerations you have for the experiment, what hypothesis test you performed ($H_0$ and $H_1$ too), and your overall conclusion/interpretation for the _business stakeholders_. Is there a follow-up you'd suggest? 

In [23]:
df.head()



Unnamed: 0                          order_id  \
0     4122928  3e6c5e89fdddcaee0eed210ec2c9cadf   
1      564306  feed6203517d3abf6aab13761633174b   
2     1872289  e33d5d7941edc281646aa37763729771   
3     3616779  7c4caa8d508fa7c3bbc25f35cdd9168a   
4     5871482  12874b29bde8bbd43fb2b95735caf9e6   

                            user_id                           page_id  \
0  90d58d967eb72656e86059ec6f208092  2fdc16a09e0016555dd4da4a3fe84414   
1  08703dab1f004eabba25aacb7f0e5484  6b0a902b9b73d5a158d0119d6feb38ac   
2  bdf1d25697e21419901c94fabdafad15  9ddb7315c4357929931b48f2b3d11c62   
3  8d2f23a732c9527d95678088a3bac122  1f86cd0bea31d54a5b511b42fd19401a   
4  5a22604f8f31ae98ee1211ece3a02004  b533f7e2003418c63fd71471264c559a   

       product site_version                 time        title  target  
0  accessories      desktop  2019-03-06 08:42:47  banner_show       0  
1     sneakers       mobile  2019-04-19 18:30:45  banner_show       0  
2      company       mobile  2019-01-20 17:20:10  banner_show       0  
3     sneakers       mobile  2019-02-20 09:38:32  banner_show       0  
4     sneakers       mobile  2019-04-24 09:19:02  banner_show       0

$H_o$: There is no difference between those purchasing sports nutrion items on mobile vs desktop.

$H_a$: There is a difference between those purchasing sports nutrition items on mobile vs desktop.

In [38]:
alpha = 0.01

In [24]:
df_AB = df[(df['time'] >= '2019-05-01') &
           (df['product'] == 'sports_nutrition')]\
.reset_index(drop=True)

In [25]:
df_AB.shape

(50508, 9)

In [26]:
df_A = df_AB[df_AB['site_version'] == 'mobile']
df_B = df_AB[df_AB['site_version'] == 'desktop']

In [27]:
df_A.head()

Unnamed: 0                          order_id  \
0     6365447  ae031dd8b5684d4a13e12f9fbca8ca6e   
2     5971896  19bda32538714d4e515e1f2209fd1298   
4     4503459  8cd4194a11ac862af37a1095ddbf9ce1   
5     8119406  109940a9c78e856ba7a10ec40615fa8f   
6     6491405  d85f7e9557b2bce18f68a1d82ecbd023   

                            user_id                           page_id  \
0  0e2135ead33f549ba2b774f6422c09b5  637a5f326a15510ca050600d47541060   
2  88c2c9e57454f3f7e088c98442f64b58  3fc1a6f8f9d1eb67ef2f242d136923c9   
4  4fa48307113761794019dda020ab3f3e  af9374cd8aef7a34346adb97fcdc3e8f   
5  c4fa641db331603f56545dc3fdfe9c0d  5ad0475859fd48494319347661159e36   
6  ab4927a8e66dfd1688e9f91d25b448ba  c1fba6ba5b10d42cc55e4a7ff147cfe3   

            product site_version                 time        title  target  
0  sports_nutrition       mobile  2019-05-20 17:39:52  banner_show       0  
2  sports_nutrition       mobile  2019-05-16 23:17:18  banner_show       0  
4  sports_nutrition       mobile  2019-05-10 08:05:57  banner_show       0  
5  sports_nutrition       mobile  2019-05-11 09:31:23  banner_show       0  
6  sports_nutrition       mobile  2019-05-05 13:40:11  banner_show       0

In [28]:
mobile_orders = sum(df_A['target'])
desktop_orders = sum(df_B['target'])

mobile_orders, desktop_orders

(303, 313)

In [29]:
total_mobile = len(df_A)
total_desktop = len(df_B)

mobile_no_orders = total_mobile - mobile_orders
desktop_no_orders = total_desktop - desktop_orders

mobile_no_orders, desktop_no_orders

(37035, 12857)

In [31]:
cont_table = np.array([(mobile_orders, mobile_no_orders),
                       (desktop_orders, desktop_no_orders)])

cont_table

array([[  303, 37035],
       [  313, 12857]])

In [33]:
result = stats.chi2_contingency(cont_table)
chi_2, p, dof, exp_table = result
result

(196.66137985526225,
 1.1179529849611601e-44,
 1,
 array([[  455.37752435, 36882.62247565],
        [  160.62247565, 13009.37752435]]))

In [35]:
mobile_CR, desktop_CR = cont_table[:, 0] / cont_table.sum(axis=1)

mobile_CR*100, desktop_CR*100

(0.8115057046440624, 2.37661351556568)

In [37]:
print(f'Conversion Rate for sports nutrition on mobile:\n\t{100*mobile_CR:.3f}%')
print(f'Conversion Rate for sports nutrition on desktop:\n\t{100*desktop_CR:.3f}%')
print('')
print(f'Absolute difference of CR: {100*(desktop_CR - mobile_CR):.3f}%')

Conversion Rate for sports nutrition on mobile:
	0.812%
Conversion Rate for sports nutrition on desktop:
	2.377%

Absolute difference of CR: 1.565%
